docs(oss): add langchain/production observability docs (#372)
langchain-ai/docs: src/oss/langchain-observability.mdx

title: Observability
---

import AlphaCallout from '/snippets/alpha-lc-callout.mdx';

<AlphaCallout />

Observability is crucial for understanding how your agents behave in production. With LangChain's `create_agent()`, you get built-in observability through [LangSmith](https://smith.langchain.com/) - a powerful platform for tracing, debugging, evaluating, and monitoring your LLM applications.

Traces capture every step your agent takes, from the initial user input to the final response, including all tool calls, model interactions, and decision points. This enables you to debug your agents, evaluate performance, and monitor usage.

## Prerequisites

Before you begin, ensure you have the following:
* A [LangSmith account](https://smith.langchain.com/) (free to sign up)

## Enable tracing

> LangChain agents created with `create_agent()` automatically support LangSmith tracing. To enable it, set the following environment variables:

Reviewer: ```suggestion
All LangChain agents automatically support LangSmith tracing. To enable it, set the following environment variables. You will receive an API key from LangSmith upon sign-up.
```

)

# Run the agent - all steps will be automatically traced
response = agent.invoke({
    "messages": [{"role": "user", "content": "Search for the latest AI news and email a summary to john@example.com"}]
})
```
:::

:::js
```ts
import { createAgent } from "@langchain/agents";

function sendEmail(to: string, subject: string, body: string): string {
    // ... email sending logic
    return `Email sent to ${to}`;
}

function searchWeb(query: string): string {
    // ... web search logic
    return `Search results for: ${query}`;
}

const agent = createAgent({
    model: "openai:gpt-4o",
    tools: [sendEmail, searchWeb],
    prompt: "You are a helpful assistant that can send emails and search the web."
});

// Run the agent - all steps will be automatically traced
const response = await agent.invoke({
    messages: [{ role: "user", content: "Search for the latest AI news and email a summary to john@example.com" }]
});
```
:::

By default, the trace will be logged to the project with the name `default`. To configure a custom project name, see [Log to a specific project](#log-to-a-specific-project).

## Trace selectively

> You can trace only specific invocations or parts of your application using LangSmith's `tracing_context` context manager:

Reviewer: ```suggestion
You may opt to trace specific invocations or parts of your application using LangSmith's `tracing_context` context manager:
```

:::python
```python
import langsmith as ls
from langchain.agents import create_agent

def send_email(to: str, subject: str, body: str):
    """Send an email to a recipient."""
    return f"Email sent to {to}"

agent = create_agent(
    model="openai:gpt-4o",
    tools=[send_email],
    prompt="You are an email assistant."
)

# This WILL be traced
with ls.tracing_context(enabled=True):
    agent.invoke({"messages": [{"role": "user", "content": "Send a test email to alice@example.com"}]})

# This will NOT be traced (if LANGSMITH_TRACING is not set)
agent.invoke({"messages": [{"role": "user", "content": "Send another email"}]})
```
:::

:::js
```ts
// TODO
```
:::

> ## Log to a specific project

Reviewer: ```suggestion
## Log to a project
```

> ### Statically

Reviewer: I think this section and dynamic could be dropdown accordions?
